internal/encryption/keyprovider/openbao/client.go

```diff
@@ -0,0 +1,88 @@
+package openbao
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+
+	openbao "github.com/openbao/openbao/api"
+)
+
+type client interface {
+	WriteWithContext(ctx context.Context, path string, data map[string]interface{}) (*openbao.Secret, error)
+}
+
+// service implements missing utility functions from openbao/api such as routing and serialization.
+type service struct {
+	c           client
+	transitPath string
+}
+
+type dataKey struct {
+	Plaintext  []byte
+	Ciphertext []byte
+}
+
+func (s service) generateDataKey(ctx context.Context, keyName string, bitSize int) (dataKey, error) {
+	path := path.Join(s.transitPath, "datakey/plaintext", url.PathEscape(keyName))
```

Comment:
I didn't notice this before, but key name must already be a path component:

```
    ^keys/(?P<name>\w(([\w-.]+)?\w)?)$
        Managed named encryption keys
```

I don't think path escaping it does anything; anything that'd need to be URL escaped wouldn't be a valid path, and URL escaping (to add `%`s) would prevent it from matching the path component too.

Reply:
I guess the one counterargument could be, if a future hypothetical update were to be created wherein either `datakey/<plaintext>/<key>` or `decrypt/<key>` had alternative mode with path suffix (e.g., `datakey/<plaintext>/<key>/<mode>` or `decrypt/<key>/<mode>`), URL encoding wouldn't make any invalid-url-component-key-names magically work, but would ensure that it gets restricted to only a single path segment. :shrug:  I don't think this type of URL scheme is likely to exist though.

Reply:
Yes, also a future hypothetical update could change the regexp of the keys potentially introducing a bug in our system so we just *a bit safer* that way.